docs/config.rst
:py:data:`readgssi.config` (essentials)
=====================================================

This module contains some things readgssi needs to operate, both command line and python-related. It contains the distribution name, author, and help text.

.. automodule:: readgssi.config
    :members:

................

* :ref:`genindex`
* :ref:`modindex`
* :ref:`search`


CHANGES.md
# Changelog

## changes since 0.0.15
- fixed [#14](https://github.com/iannesbitt/readgssi/issues/14) which incorrectly asserted that user-set samples per meter was negative when it was not (sign was flipped) but processed the file correctly regardless
- moved zoom calculation and automatic file naming to their own functions
- added sphinx autodoc descriptions to some functions, in order to work towards solving [#9](https://github.com/iannesbitt/readgssi/issues/9)
- added comments to datetime parsing function
- fixed sampling frequency calculations to the value of `epsr` set in the header, which will prevent perceived frequency wander in frequency-based filters if the user sets `epsr`
- changed the behavior of the readdzt script to automatically search for a DZG file of the same name, and load that data if possible. this should allow python users to read all three variables (header, arrays, gps) if GPS exists.
- fixed command line control of DPI
- switched to "gray" colormap from "Greys". "Greys" rendered opposite polarity [#18](https://github.com/iannesbitt/readgssi/issues/18)

## changes since 0.0.14
- added support for the 5103 400 MHz antenna, and an antenna called `CUSTOM` in the header, for which center frequency is not known but is likely adjustable
- added support for the 350 HS hyperstacking antenna
- fixed a bug that created an extra matplotlib axis
- added rounding to epsilon_r value on y-axis depth-type display
- added a patch for [#12](https://github.com/iannesbitt/readgssi/issues/12) to improve axis sizing and tight_layout() handling
- changed background removal (BGR) function to operate similar to RADAN boxcar function
  - this change means that you can no longer simply specify `-r` for full-width background removal. new nomenclature for full-width BGR flag is `-r 0`. `-r 100` will apply a 100-trace rolling boxcar BGR (after normalization and stacking)
  - this is done with the `scipy.ndimage.filters.uniform_filter1d()` function and means that `scipy` is now explicitly required
  - function will automatically filter out low-frequency washout noise caused by bgr, using obspy's implementation of scipy's butterworth bandpass from `0.1*f` to `10*f`, where `f` is antenna frequency. nyquist is automatically accounted for by obspy.
  - BGR value is displayed in title if set > 0, or as "full" if BGR is full-width (i.e. `-r 0`)
- added zerophase vertical triangular FIR filter, and changed default filter from butterworth to triangular to attempt to match that in RADAN. butterworth is no longer used (but kept for posterity).
  - this function uses the `scipy.signal.firwin()` filter design and `scipy.signal.lfilter()` implementation
- added ability to suppress plot title using `-T` or `--titleoff` from the command line or title=False in main python function
- changed `ns_per_zsample` in header to read data shape from header rather than array size
- added a bug fix for filtering that caused the sample rate to be set too high
- significantly improved triangular bandpass filter, so that it nearly matches RADAN
- when Z axis value is set to "samples", the zeroed samples are taken into account. so, if the array had 150 samples sliced off, the plot axis will now start at 150 instead of 0
- added per-channel handling of timezero. you can pass a list with `-Z 233,150,92` (or `zero=[233,150,92,None]` in python). defaults to `-Z 2` (`zero=[2,None,None,None]`)
- added programmable zoom with `-e left,right,up,down` flag and list
- cleaned up some dangling floats
- updated help text and README help text
- added automatic tight bbox axes for savefig
- changed from 150 to 300 dpi for easier poster printing measurements
- fixed calculation for two-way time (previously per [#17](https://github.com/iannesbitt/readgssi/issues/17), calculation erroneously displayed one-way time)

## changes since 0.0.13
- updated examples in README
- fixed [#16](https://github.com/iannesbitt/readgssi/issues/16) which incorrectly applied GPS array values for normalization

## changes since 0.0.12
- fixed [#13](https://github.com/iannesbitt/readgssi/issues/13) by adding `SS MINI` to list of recognized antenna codes
- fixed bug ([#15](https://github.com/iannesbitt/readgssi/issues/15)) which caused runaway memory usage when distance-normalizing large files. files are now distance normalized in chunks.
- fixed a bug that caused a read failure on DZG files with RMC strings before GGA (iat function didn't like my slicing)

## changes since 0.0.11
- fixed bug that prevented reading NMEA RMC-only DZG files. now fully compatible with DZGs created with [gpx2dzg](https://github.com/iannesbitt/gpx2dzg)
- added `-d` flag documentation to help text and readme
- updated docstring formatting
- added PyPI downloads badge
- added ability to manually set samples per meter using `-d <float>` flag
- added support and badge for travis-ci in anticipation of test suite
- `-V` or `--version` outputs version string
- fixed bug ([#10](https://github.com/iannesbitt/readgssi/issues/10)) which read NMEA GGA strings incorrectly from DZGs that don't contain NMEA RMC
- fixed bug ([#11](https://github.com/iannesbitt/readgssi/issues/11)) which prevented SIR-20 files with antenna code `GSSI` from being read correctly

## changes since 0.0.10
- addresses [#7](https://github.com/iannesbitt/readgssi/issues/7) to add support for DZT files created with gprMax 
- added explicit support several antenna codes specific to SIR-3000 units and before
- added the ability to label radargram axes with distance and time units
- created readthedocs support files in anticipation of more full documentation and a future stable version release

## changes since 0.0.9
- a patch was added that includes several important bug fixes
  - added support for the `400MHz` antenna type
  - fixed a bug that affected the stacking routine
  - fixed a bug with the time zero variable not being available when implementing the naming scheme

readgssi/config.py
from datetime import datetime
from readgssi.__init__ import __version__, name
import pkg_resources

"""
This module contains some things readgssi needs to operate, both command line and python-related.
"""

dist = pkg_resources.get_distribution(name)
year = datetime.now().year
author = 'Ian Nesbitt'
affil = 'School of Earth and Climate Sciences, University of Maine'

help_text = u'''Help text:
############################################################
 readgssi version %s

 Copyleft %s %s 2017-%s
 %s
############################################################

usage:
readgssi -i input.DZT [OPTIONS]

required flags:
     OPTION     |      ARGUMENT       |       FUNCTIONALITY
-i, --input     | file:  /dir/f.DZT   |  input DZT file

optional flags:
     OPTION     |      ARGUMENT       |       FUNCTIONALITY
-o, --output    | file:  /dir/f.ext   |  output file. if not set, will be named similar to input
-f, --format    | string, eg. "csv"   |  output format (csv is the only working format currently)
-p, --plot      | +integer or "auto"  |  plot size. will be x inches high or "auto". default: 10. see also -D to set DPI
-D, --dpi       | positive integer    |  set the plot DPI for figure making. defaults to 150
-T, --titleoff  |                     |  turn the plot title off (useful for figure making)
-x, --xscale    | string, eg. "dist"  |  x units. will attempt to convert the x-axis to distance, time, or trace units based on header values
-z, --zscale    | string, eg. "time"  |  z units. attempt to convert the x-axis to depth, time, or sample units based on header values
-e, --zoom      | list of +int [LRUD] |  set a zoom to automatically jump to. list order is [left,right,up,down] and units are the same as axis
-n, --noshow    |                     |  suppress matplotlib popup window and simply save a figure (useful for multi-file processing)
-c, --colormap  | string, eg. "Greys" |  specify the colormap (https://matplotlib.org/users/colormaps.html#grayscale-conversion)
-g, --gain      | positive float      |  gain constant (higher=greater contrast, default: 1)
-r, --bgr       | +integer or zero    |  horizontal background removal (useful to remove ringing). zero=full width; positive=window size (after stacking)
-R, --reverse   |                     |  reverse (flip array horizontally)
-w, --dewow     |                     |  trinomial dewow algorithm
-t, --bandpass  | +int-+int (MHz)     |  triangular FIR bandpass filter applied vertically (positive integer range in megahertz; ex. 70-130)
-b, --colorbar  |                     |  add a colorbar to the radar figure
-a, --antfreq   | positive integer    |  set antenna frequency. overrides header value
-s, --stack     | +integer or "auto"  |  set trace stacking value or "auto" to autostack to ~2.5:1 x:y axis ratio
-N, --normalize |                     |  distance normalize; reads .DZG NMEA data file if it exists; otherwise tries to read CSV with lat, lon, and time fields
-d, --spm       | positive float      |  specify the samples per meter (spm). overrides header value
-m, --histogram |                     |  produce a histogram of data values
-E, --epsr      | float > 1.0         |  user-defined epsilon sub r (sometimes referred to as "dielectric") if set, ignores value in DZT header
-Z, --zero      | +int or list of int |  timezero: skip samples before direct wave. samples are removed from the top of the trace. use list for multi-channel

naming scheme for exports:
  CHARACTERS    |      MEANING
    Ch0         |  Profile from channel 0 (can range from 0 - 3)
    Dn          |  Distance normalization
    Tz233       |  Time zero at 233 samples
    S8          |  Stacked 8 times
    Rv          |  Profile read in reverse (flipped horizontally)
    Bgr75       |  Background removal filter with window size of 75
    Dw          |  Dewow filter
    Bp70-130    |  triangular FIR filter applied from 70 to 130 MHz
    G30         |  30x contrast gain
    Z10.20.7.5  |  zoom from 10-20 axis units on the x-axis and 5-7 on the z-axis
''' % (__version__, u'\U0001F12F', author, year, affil)

version_text = '%s %s' % (name, __version__)